paymentsheet/src/main/java/com/stripe/android/paymentsheet/ui/PrimaryButtonContainerFragment.kt

```diff
@@ -0,0 +1,177 @@
+import com.stripe.android.paymentsheet.PaymentOptionsActivity
+import com.stripe.android.paymentsheet.PaymentOptionsViewModel
+import com.stripe.android.paymentsheet.PaymentSheetActivity
+import com.stripe.android.paymentsheet.PaymentSheetViewModel
+import com.stripe.android.paymentsheet.R
+import com.stripe.android.paymentsheet.databinding.FragmentPrimaryButtonContainerBinding
+import com.stripe.android.paymentsheet.model.PaymentSheetViewState
+import com.stripe.android.paymentsheet.navigation.PaymentSheetScreen
+import com.stripe.android.paymentsheet.utils.launchAndCollectIn
+import com.stripe.android.paymentsheet.viewmodels.BaseSheetViewModel
+import com.stripe.android.uicore.StripeTheme
+import com.stripe.android.uicore.getBackgroundColor
+
+internal abstract class BasePrimaryButtonContainerFragment : Fragment() {
+
+    protected var viewBinding: FragmentPrimaryButtonContainerBinding? = null
+
+    abstract val viewModel: BaseSheetViewModel
+
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?
+    ): View? {
+        viewBinding = FragmentPrimaryButtonContainerBinding.inflate(inflater, container, false)
+        return viewBinding?.root
+    }
+
+    abstract fun resetPrimaryButtonState()
+
+    protected fun setupPrimaryButton() {
+        val viewBinding = viewBinding ?: return
+
+        viewModel.primaryButtonUIState.launchAndCollectIn(viewLifecycleOwner) { state ->
+            state?.let {
+                viewBinding.primaryButton.setOnClickListener {
+                    state.onClick?.invoke()
+                }
+                viewBinding.primaryButton.setLabel(state.label)
+                viewBinding.primaryButton.isVisible = state.visible
+            } ?: run {
+                resetPrimaryButtonState()
+            }
+        }
+
+        viewModel.amount.launchAndCollectIn(viewLifecycleOwner) {
+            resetPrimaryButtonState()
+        }
+
+        viewModel.selection.launchAndCollectIn(viewLifecycleOwner) {
+            resetPrimaryButtonState()
+        }
+
+        viewModel.isPrimaryButtonEnabled.launchAndCollectIn(viewLifecycleOwner) { isEnabled ->
+            viewBinding.primaryButton.isEnabled = isEnabled
+        }
+
+        viewBinding.primaryButton.setAppearanceConfiguration(
+            StripeTheme.primaryButtonStyle,
+            tintList = viewModel.config?.primaryButtonColor ?: ColorStateList.valueOf(
+                StripeTheme.primaryButtonStyle.getBackgroundColor(requireActivity().baseContext)
+            )
+        )
+    }
+
+    override fun onDestroyView() {
+        viewBinding = null
+        super.onDestroyView()
+    }
+}
+
+internal class PaymentSheetPrimaryButtonContainerFragment : BasePrimaryButtonContainerFragment() {
+
+    override val viewModel: PaymentSheetViewModel by activityViewModels {
+        PaymentSheetViewModel.Factory {
+            requireNotNull(
```

Comment:
If the view model isn't already created, we're in an unexpected state. Do we ever test this factory? Should we just throw instead?

Reply:
Changed it as discussed via Slack.